### PR TITLE
Update `FocusManager` platform check to include iOS

### LINKED ISSUE #148475: Autofill Hints not populating in text fields on iOS with Flutter 3.22.0
### Steps to reproduce

1. Update Flutter to version 3.22.0.
2. Ensure .well-known/apple-app-site-association is correctly configured on the server.
3. Implement autofill hints in your Flutter app.
4. Test the autofill functionality on an iOS device.
5. Try to autofill with some saved credentials

### Expected results

Input fields should be automatically populated with relevant data when using autofill on iOS, similar to how it works on Android.

### Actual results

Autofill suggestions appear on iOS, but selecting a suggestion does not populate the input fields.

### Code sample

```dart
 AutofillGroup(
      child: Scaffold(
        body: GoldRippleGradientBackground(
          child: CustomScrollView(slivers: <Widget>[
            SliverFillRemaining(
              hasScrollBody: false,
              child: SafeArea(
                child: Padding(
                  padding:
                      const EdgeInsets.all(Dimensions.defaultScreenPadding),
                  child: Form(
                    key: formKey,
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      mainAxisAlignment: MainAxisAlignment.center,
                      children: <Widget>[
                        const Spacer(),
                          TextFormField(
                            key: const Key('email_sign_in'),
                            autofillHints: const <String>[
                              AutofillHints.email,
                              AutofillHints.username
                            ],
                            decoration: InputDecoration(
                              labelText: context.strings.authEmailLabel,
                            )),
                          TextFormField(
                              key: const Key('password_sign_in'),
                              autofillHints: const <String>[
                                AutofillHints.password,
                              ],
                              decoration: InputDecoration(
                                labelText: context.strings.authPasswordLabel,
                          )),                      
                      ],
                    ),
                  ),
                ),
              ),
            ),
          ]),
        ),
      ),
    );
```

### Screenshots or Video

https://github.com/flutter/flutter/assets/57009558/382d0e3a-840f-43a0-a758-24e63ae086dc

### Logs

_No response_

### Flutter Doctor output

[✓] Flutter (Channel stable, 3.22.0, on macOS 14.4.1)
    • Flutter version 3.22.0 on channel stable at 
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 5dcb86f68f (7 days ago), 2024-05-09 07:39:20 -0500
    • Engine revision f6344b75dc
    • Dart version 3.4.0
    • DevTools version 2.34.3

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK  path
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = path
    • Java binary at: path
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11572160)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.4)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15F31d
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = /Applications/Brave Browser.app/Contents/MacOS/Brave Browser

[✓] Android Studio (version 2023.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11572160)

[✓] Network resources
    • All expected network resources are available.

### PR DESCRIPTION
**Pull request by nate-thegrate**
Both iOS and Android run into issues when the FocusManager starts responding to app lifecycle changes.

Fortunately, this feature is primarily meant for desktop platforms, so the problem can be resolved with a platform check.

fixes https://github.com/flutter/flutter/issues/148475

### CODE REVIEW

**Comment by nate-thegrate** [comment]
Currently, the test `'flutter create should tool exit if the template manifest cannot be read'` is failing (on multiple branches, not just this one).

My guess is that `expectLater` has a bug where it doesn't actually `await` the futures like it should, so the test only succeeds when the future finishes quickly enough.

File: packages/flutter/lib/src/widgets/focus_manager.dart [hunk 1]
```diff
@@ -1521,12 +1521,21 @@ class FocusManager with DiagnosticableTreeMixin, ChangeNotifier {
     if (kFlutterMemoryAllocationsEnabled) {
       ChangeNotifier.maybeDispatchObjectCreation(this);
     }
-    if (kIsWeb || defaultTargetPlatform != TargetPlatform.android) {
+    final bool shouldListenToAppLifecycle = switch (defaultTargetPlatform) {
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/widgets/focus_manager.dart:1524]: Having the focus system automatically handle app lifecycle state changes is convenient in most cases, but what if a UI component wants to respond to app lifecycle events and user-initiated focus changes differently? If we do this automatically in the focus system, there's no way for a component to distinguish user-initiated focus changes and app-lifecycle focus changes.

For example a text field establishes / tears down an input connection based on whether it currently has focus.  Automatically unfocusing on app pause doesn't work well. It makes more sense to me that UI components should have logic to handle app lifecycle changes themselves (at least for text fields that's the case).

Comment (nate-thegrate) [reply packages/flutter/lib/src/widgets/focus_manager.dart:1524]: Setting all UI components (not just text fields) to automatically unfocus/refocus based on app lifecycle state matches the behavior of other non-Flutter apps.

<details open> <summary>Example (Google Chrome)</summary>

chrome focus changes

</details>

But the fact that a widget may want to differentiate between app lifecycle events and user-initiated focus changes is a great point.

In my mind, it makes sense to apply the iOS platform check today, and then sometime soon we should swap out the current app lifecycle listener for a more robust implementation that allows for the `TextField` input connection optimization you mentioned.

**Review by LongCatIsLooong — APPROVED** [review]

**Comment by polina-c** [comment]
FYI: this PR introduced a memory leak

https://github.com/flutter/flutter/issues/148985

**Comment by Para123123** [comment]
Flutter 3.32.5 also ran into this issue

**Pull request merged**